### PR TITLE
MB-2114 Adds prime api support end point for updating an MTO shipment's status

### PR DESCRIPTION
**Pull request by sandy-wright**
## Description

- Adds `/mto-shipments/:mtoShipmentID/status` to the prime API (in `support.yaml`)
- Support endpoint only - will not be used in production
- Useful for internal testing of the prime API

## Reviewer Notes

The path has been edited from the original ghc API endpoint to exclude the `moveTaskOrderID`. This value was never used and was only extra overhead for testing. Functionality is otherwise the same as the ghc endpoint.

## Setup

Should be testable with Postman or the `prime-api-client` CLI following [these instructions](https://github.com/transcom/mymove/wiki/run-prime-docker). Sample command:

```sh
go run ./cmd/prime-api-client support-patch-mto-shipment-status --etag <etag> --filename <filename> --insecure
```

Make sure you have run the following commands prior to testing:

```sh
$ make db_dev_e2e_populate
$ make server_run
```

### Testing with Postman
If Postman is not set up for mutual TLS, please follow instructions here: https://github.com/transcom/mymove/wiki/setup-postman-to-make-mutual-tls-api-calls

You'll need to go through these steps for the prime support environment:

**Prime Local Environment**
Host: primelocal
baseUrl: https://primelocal:9443
prime: /prime/v1
support: /support/v1

#### Get MTO Shipments
You'll need to get a shipment's id and eTag before you can update its status. 

- Use the GET: "Gets all move task orders" `{{baseUrl}}{{prime}}/move-task-orders?since=<since>` endpoint to return MTOs and their shipments. 
- Choose a shipment ID and eTag from this output. 

#### Update status
In Postman, using the same prime support environment:

- Send a PATCH request to `{{baseUrl}}{{support}}/mto-shipments/:mtoShipmentID/status`
- Pass in the id as the `mtoShipmentID` parameter
- Pass in the eTag as the value for `If-Match` in the Headers
- Pass an object in the Body tag containing a status and rejection reason. Example:

```json
{
    "status": "REJECTED",
    "rejectionReason": "Not valid"
}
```

You should get a JSON response with all of the shipment fields if the update was successful, otherwise you should see a reasonably informative error message.

#### Get MTO Shipments
Use the command:

```sh
go run ./cmd/prime-api-client --insecure fetch-mtos | jq
```

to receive a list of MTOs and their shipments, just like in Postman. Grab a shipment ID and its eTag.

#### Update status
Create test JSON file in the following format:

```json
{
    "id": "36e4da9e-0512-4038-8e29-acd4ad7a29d6",
    "status": "REJECTED",
    "rejectionReason": "oh no no",
}
```

Run command:

```sh
go run ./cmd/prime-api-client support-patch-mto-shipment-status --etag <etag> --filename <filename> --insecure | jq
```

If the action was a success, you should see a similar response print out:

```json
{
  "approvedDate": "0001-01-01",
  "createdAt": "2020-04-17T16:15:49.509Z",
  "customerRemarks": "please treat gently",
  "destinationAddress": {
    "city": "Fairfield",
    "country": "US",
    "eTag": "MjAyMC0wNC0xN1QxNjoxNTo0OS40OTkyMzha",
    "id": "750dfdcd-61e1-48d0-8049-5413db24f974",
    "postal_code": "94535",
    "state": "CA",
    "street_address_1": "987 Any Avenue",
    "street_address_2": "P.O. Box 9876",
    "street_address_3": "c/o Some Person"
  },
  "eTag": "MjAyMC0wNC0xN1QxODozNDo0OS4yNjk0MzNa",
  "id": "36e4da9e-0512-4038-8e29-acd4ad7a29d6",
  "moveTaskOrderID": "19c0f3f2-7ba8-4789-af21-9634e2d3ce5c",
  "pickupAddress": {
    "city": "Beverly Hills",
    "country": "US",
    "eTag": "MjAyMC0wNC0xN1QxNjoxNTo0OS40OTU2NDla",
    "id": "eb851067-a8b3-4d49-90e2-a10c88554604",
    "postal_code": "90210",
    "state": "CA",
    "street_address_1": "123 Any Street",
    "street_address_2": "P.O. Box 12345",
    "street_address_3": "c/o Some Person"
  },
  "rejectionReason": "oh no no",
  "requestedPickupDate": "2018-03-15",
  "scheduledPickupDate": "0001-01-01",
  "secondaryDeliveryAddress": {
    "city": "Beverly Hills",
    "country": "US",
    "eTag": "MjAyMC0wNC0xN1QxNjoxNTo0OS41MDYwOTda",
    "id": "1b034b62-dd6e-4afe-bd64-f79f6d25416c",
    "postal_code": "90210",
    "state": "CA",
    "street_address_1": "123 Any Street",
    "street_address_2": "P.O. Box 12345",
    "street_address_3": "c/o Some Person"
  },
  "secondaryPickupAddress": {
    "city": "Beverly Hills",
    "country": "US",
    "eTag": "MjAyMC0wNC0xN1QxNjoxNTo0OS41MDI3MTRa",
    "id": "922811e7-345e-47bb-9c1e-e5b2c61fc9bf",
    "postal_code": "90210",
    "state": "CA",
    "street_address_1": "123 Any Street",
    "street_address_2": "P.O. Box 12345",
    "street_address_3": "c/o Some Person"
  },
  "shipmentType": "HHG",
  "status": "REJECTED",
  "updatedAt": "2020-04-17T18:34:49.269Z"
}
```

Otherwise, try:

```sh
go run ./cmd/prime-api-client support-patch-mto-shipment-status --etag <etag> --filename <filename> --insecure -v
```

to see the detailed error response.

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/tree/master/docs/backend.md#logging)
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-2114) for this change

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -12589,10 +12589,10 @@ react-color@^2.17.0:
     reactcss "^1.2.0"
     tinycolor2 "^1.4.1"
 
-react-day-picker@=7.4.4:
-  version "7.4.4"
-  resolved "https://registry.yarnpkg.com/react-day-picker/-/react-day-picker-7.4.4.tgz#666b9ef3845b430d30407c808cd034180f123582"
-  integrity sha512-jGhgGUqkY2i5tOQ8/XhwPFuXXwBDq4KxQaQ2XQh4/RnNFaLkUjBOTfA/WbEHkmNk/JKdE0fNxVRzVwxnYdckww==
+react-day-picker@=7.4.5:
```
Comment (shimonacarvalho) [inline yarn.lock:12592]: Did you mean to include this file?

File: swagger/support.yaml [hunk 2]
```diff
@@ -67,6 +67,63 @@ paths:
         - contracting_officer
         - ppm_office_users
       summary: Change the status of a move task order to make it available to prime
+  '/mto-shipments/{shipmentID}/status':
```
Comment (shimonacarvalho) [inline swagger/support.yaml:70]: Upto you, but you could go ahead and flesh out the descriptions here. Rather than a second ticket.

File: go.sum [hunk 3]
```diff
@@ -28,8 +28,8 @@ github.com/asaskevich/govalidator v0.0.0-20190424111038-f61b66f89f4a h1:idn718Q4
 github.com/asaskevich/govalidator v0.0.0-20190424111038-f61b66f89f4a/go.mod h1:lB+ZfQJz7igIIfQNfa7Ml4HSf2uFQQRzpGGRXenZAgY=
 github.com/asaskevich/govalidator v0.0.0-20200108200545-475eaeb16496 h1:zV3ejI06GQ59hwDQAvmK1qxOQGB3WuVTRoY0okPTAv0=
 github.com/asaskevich/govalidator v0.0.0-20200108200545-475eaeb16496/go.mod h1:oGkLhpf+kjZl6xBf758TQhh5XrAeiJv/7FRz/2spLIg=
-github.com/aws/aws-sdk-go v1.30.7 h1:IaXfqtioP6p9SFAnNfsqdNczbR5UNbYqvcZUSsCAdTY=
-github.com/aws/aws-sdk-go v1.30.7/go.mod h1:5zCpMtNQVjRREroY7sYe8lOMRSxkhG6MZveU8YkpAk0=
+github.com/aws/aws-sdk-go v1.30.8 h1:4BHbh8K3qKmcnAgToZ2LShldRF9inoqIBccpCLNCy3I=
```
Comment (shimonacarvalho) [inline go.sum:31]: did you mean to check in this file?

File: go.mod [hunk 4]
```diff
@@ -6,7 +6,7 @@ require (
 	github.com/0xAX/notificator v0.0.0-20191016112426-3962a5ea8da1 // indirect
 	github.com/99designs/aws-vault v4.5.1+incompatible
 	github.com/99designs/keyring v1.1.4
-	github.com/aws/aws-sdk-go v1.30.7
+	github.com/aws/aws-sdk-go v1.30.8
```
Comment (shimonacarvalho) [inline go.mod:9]: did you mean to check in this file?

File: cmd/prime-api-client/main.go [hunk 5]
```diff
@@ -103,6 +103,26 @@ func main() {
 	initUpdateMTOStatusFlags(makeAvailableToPrimeCommand.Flags())
 	root.AddCommand(makeAvailableToPrimeCommand)
 
+	createPaymentRequestCommand := &cobra.Command{
```
[the remark this answers is not in the document]
Comment (sandy-wright) [reply cmd/prime-api-client/main.go:106]: yes! I did a rebase on master to resolve conflicts, but now it looks like work is duplicated. a merge might fix it? if you compare the files with master, there aren't any differences. I don't know why the files are registering in the diff

File: pkg/handlers/supportapi/mto_shipment.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+	if err != nil {
+		logger.Error("UpdateMTOShipmentStatus error: ", zap.Error(err))
+
+		switch e := err.(type) {
+		case services.NotFoundError:
+			notFoundPayload := &supportmessages.Error{Message: handlers.FmtString(fmt.Sprintf("id: '%s' not found", shipmentID.String()))}
+			return mtoshipmentops.NewPatchMTOShipmentStatusNotFound().WithPayload(notFoundPayload)
+		case services.InvalidInputError:
+			payload := payloadForValidationError("Validation errors", "UpdateShipmentMTOStatus", h.GetTraceID(), e.ValidationErrors)
```
Comment (shimonacarvalho) [inline pkg/handlers/supportapi/mto_shipment.go:65]: You want the first param title to be a predefined value, and the second param to be an explanation of the error. LIke so, 
`		payload := payloadForValidationError(handlers.ValidationErrMessage, "The input provided did not pass validation.", h.GetTraceID(), verrs)
`

You could put a more description detail message than this example, if you know more, but the verrs should provide more detail.

**Comment by sandy-wright** [comment]
aha, so the target branch state in github doesn't update, so I rebased and it wasn't seeing that master already had those commits. If I toggle the base to another branch and then back to master, it updates again. (sorry Lindsay if you got a notification from that, I randomly picked your branch for that brief moment! I'll only use my own branches from now on)

You can also get to a true diff using this url (for any two branches): https://github.com/transcom/mymove/compare/master...sw-mb-2114-support-endpoint-shipment-status

File: pkg/handlers/supportapi/mto_shipment.go (hunk 6 above)
Comment (LeDeep) [reply pkg/handlers/supportapi/mto_shipment.go:65]: if you can make this one change; to use the errors handler, as Shimona suggests, then this should be ready for acceptance

**Comment by sandy-wright** [comment]
okay, I think I've addressed everything I needed to in this PR! @shimonacarvalho - if you would be able to check my tests again, I would really appreciate it. let me know if there's anything else I need to modify!

there is a circleci failure as well, but it seems to be related to django? so I'll try updating with master and see if that fixes it in the meantime

**Pull request merged**